### DISCUSSION TITLE
Applying `semver.stage=snapshot` between release condidates

### DISCUSSION THREAD

**Discussion by bric3** [JavierSegoviaCordoba/semver-gradle-plugin#91] [Q&A] [answered]
Hi, there,

Thank you for the plugin and the work. I'm a bit unsatisfied by the state of other plugins, I have been using some for years, but I like the simpleness of yours.

I have one question about a behavior I'm not sure is currently possible. I making a library that is in the release candidate stage, but in-between the RCs I'd like to publish snapshots, because OSSRH repo for snapshots, expects versions to end with `-SNAPSHOT`.

Currently the git graph looks like, version `0.0.1` is not yet released, only few candidates :

```mermaid
gitGraph
    commit
    commit id: "fix(osgi): Remove split package to fix OSGI ClassLoading error" tag: "v0.0.1-rc.3"
    commit
    commit id: "around 50 commits"
    commit
    commit id: "last commit" type: HIGHLIGHT
```

But the command `printSemver`, issues the following

```shell
$ ./gradlew -Psemver.stage=snapshot printSemver
Type-safe project accessors is an incubating feature.

> Task :printSemver
semver for fireplace: 0.1.1-SNAPSHOT
```

I expected it to be `0.0.1-SNAPSHOT` as _snapshots_ seems like a smaller increment that _alphas_, _betas_, _candidates_. I believe snapshots should only increment after a _final release_.

**Comment by bric3** [comment]
Maybe out of scope there, I can create issue if needed.

----

I struggled a bit to make the `-Psemver.stage=snapshot` property work, because I was using the `tagPrefix` on my root project as I was thinking I needed to tell the plugin that git tags are prefixed by `v`.

```kotlin
plugins {
  id("com.javiersc.semver.gradle.plugin") version "0.3.0-alpha.5"
}

semver {
    tagPrefix.set("v")
}
```

So the output was on a clean repo

```shell
$ ./gradlew -Psemver.stage=snapshot printSemver
Type-safe project accessors is an incubating feature.

> Task :printSemver
semver for fireplace: v0.0.1-rc.3.54+e333201
```

While debugging the plugin I found the reason is related to whether tag prefixes are equal

https://github.com/JavierSegoviaCordoba/semver-gradle-plugin/blob/72b1941a9cc178243cb14e38d6a9f21c2cfeb548/semver-gradle-plugin/main/kotlin/com/javiersc/semver/gradle/plugin/valuesources/VersionValueSource.kt#L42-L50

This makes me wonder how to use properly in a multi-project where all sub-projects have the same version. Am I using this properly ?

----

That reminds I have some logic whether the version is a snapshot or a _releasable_ artefact ; this could be useful if the plugin exposed some properties like `isDirty`, or `isSnaphsot`, or the current stage.

**Reply by bric3** [reply]
OK, after a little more debugging the two issues seem related, i.e. the snapshot incrementing to 0.1.1, is because the git tag is not read if the tag prefix is not set.

So one need to set

```properties
semver.tagPrefix=v
semver.project.tagPrefix=v
```

But then the default snapshot increment is not what I'm looking for, i'd like no increment. In other word only increment snapshot if the last tag is a _final release_.

```shell
$ ./gradlew -Psemver.stage=snapshot -Psemver.checkClean=false printSemver
Type-safe project accessors is an incubating feature.

> Task :printSemver
semver for fireplace: v0.0.2-SNAPSHOT
```

**Reply by JavierSegoviaCordoba** [reply]
You should use the same `tagPrefix` for multi-projects that shares the version. The default tag is `""` (empty `String`) so if you want to use `v` you need to set the version via the extension or via `gradle.properties`. 

The `gradle.properties` solution is added for incompatibilities with other plugins that are not respecting Gradle lazy configuration and are picking the version in the configuration phase.

**Reply by bric3** [reply]
oh so you mean something like this (and without the properties) ?

```
allprojects {
    apply(plugin = "com.javiersc.semver.gradle.plugin")
    semver {
        tagPrefix.set("v")
    }
}
```

**Reply by JavierSegoviaCordoba** [reply]
Yeah, but you need to indicate to the plugin with which task are you working.

All projects use `v` prefix with that configuration, if you run `./gradlew printSemver "-Psemver.stage=rc"`, it will not work as you are passing implicitly the empty string, you should indicate via CLI which tag are you going to work `./gradlew printSemver "-Psemver.stage=rc" "-Psemver.tagPrefix=v"`

But as you are not going to have different tags per project, it is easy what you are doing, which is setting `semver.tagPrefix=v` in a `gradle.properties` file.

The `semver.project.tagPrefix=v` is for compatibility issues, but you can keep it too as it is safer as some plugins are not respecting the configuration phase correctly.

**Reply by bric3** [reply]
Oh OK, I need both the version in the gradle file and in the properties file. So I now have this

| build.gradle.kts | gradle.properties |
| ---------------- | ----------------- |
| <pre lang="kotlin">semver {&#13;    tagPrefix.set("v")&#13;}<pre> | <pre lang="properties">semver.tagPrefix=v</pre>


and the next version is printed accordingly to the stage e.g. snapshot : `v0.0.2-SNAPSHOT` (i.e. with the other mentioned issue).

**Comment by JavierSegoviaCordoba** [comment]
Hi @bric3!

I think I set the default version to `0.1.0` when there is no tag in the branch, so probably `0.0.1-rc.1` is caused by a bug when there is no tag.

`-SNAPSHOT` versions AFAIK are treated differently in Gradle. Based on semantic versioning, `1.2.3-SNAPSHOT` is higher than `1.2.3-rc.1` as `S` is higher than `R`, but in the [Gradle version ordering](https://docs.gradle.org/current/userguide/single_versions.html#version_ordering) it isn't, as `rc` is considered higher than `SNAPSHOT`. I added this rule specifically because there is a special repository for snapshots, but I should add the rest of the rules too (like `dev` one).

Tools like Dependabot, Renovate, third-party plugins like [Ben Manes one](https://github.com/ben-manes/gradle-versions-plugin), or a future Gradle official plugin allow updating libraries automatically respecting those rules.

So supposing you are publishing the version `1.2.3-rc.1`, and with each new commit, you want to publish the typical `SNAPSHOT` which is indeed ahead of `1.2.3-rc.1`, if it is published with the same core, `1.2.3-SNAPSHOT`, it will considered lower than `1.2.3-rc.1` which is not true, that is the reason the plugin set the version to `1.2.4-SNAPSHOT`.

I could add a flag to toggle this behavior, but I would like to know your opinion about the mentioned problem before adding it.

**Reply by bric3** [reply]
Hi @JavierSegoviaCordoba 

Thanks for replying this fast, sorry I got caught in other things yesterday. So you are bringing good points, however I believe this statement is erroneous.

> but in the [Gradle version ordering](https://docs.gradle.org/current/userguide/single_versions.html#version_ordering) it isn't, as rc is considered higher than SNAPSHOT

Typically [this point][4] lead me to believe snapshots are actually higher than RCs

> * The strings `rc`, `snapshot`, `final`, `ga`, `release` and `sp` are considered **higher** than any other string part (sorted in this order): `1.0-zeta` < `1.0-rc` < `1.0-snapshot` < `1.0-final` < `1.0-ga` < `1.0-release` < `1.0-sp` < `1.0`.

Moreover this only applies when the consuming project defines [dynamic versions][3] (so something like `0.+`). If the version is specifically defined, then Gradle should use the specified use the right version._

That said I'm not sure of the implication for other tools like renovate or the gradle versions plugin. I believe renovate might suggest a SNAPSHOT, if the current version is a RC, but that's somewhat expected, and there might be mechanisms to filter out `SNAPSHOT`s if they are not wanted in the consuming project.

> I added this rule specifically because there is a special repository for snapshots

I'm not sure I understand what you mean. Here's what I'm facing when publishing a library over the OSSRH repositories, with snapshots repos for snapshots and staging repos for publishing actual releases that should be available on central.

With this build script, Gradle will obviously pick the snapshot, the [snapshot repo][1] doesn't have any _candidates_ (on [central][2] only `0.0.1-rc2` and `0.0.1-rc3` are published).

<details><summary>explicit version example</summary>

```kotlin
plugins {
  `java`
}

dependencies {
  implementation("io.github.bric3.fireplace:fireplace-swing:0.0.1-SNAPSHOT")
}

repositories {
  mavenCentral()
  maven("https://s01.oss.sonatype.org/content/repositories/snapshots")
}
```

And the `dependencyInsight` task of course confirm that gradle is picking the snapshot.

```shell
$ gradle dependencyInsight --configuration=runtimeClasspath --dependency fireplace

> Task :dependencyInsight
io.github.bric3.fireplace:fireplace-swing:0.0.1-SNAPSHOT:20221222.093112-36
  Variant runtimeElements:
    | Attribute Name                 | Provided     | Requested    |
    |--------------------------------|--------------|--------------|
    | org.gradle.status              | integration  |              |
    | org.gradle.category            | library      | library      |
    | org.gradle.dependency.bundling | external     | external     |
    | org.gradle.jvm.version         | 11           | 19           |
    | org.gradle.libraryelements     | jar          | jar          |
    | org.gradle.usage               | java-runtime | java-runtime |
    | org.gradle.jvm.environment     |              | standard-jvm |

io.github.bric3.fireplace:fireplace-swing:0.0.1-SNAPSHOT
\--- runtimeClasspath
```

Using `latest.integration` instead returns the latest snapshot (`0.0.1-SNAPSHOT`), and `latest.release` returns `0.0.1-rc3`

</details>

  [1]: https://s01.oss.sonatype.org/content/repositories/snapshots/io/github/bric3/fireplace/fireplace-swing/
  [2]: https://repo1.maven.org/maven2/io/github/bric3/fireplace/fireplace-swing/0.0.1-rc3/
  [3]: https://docs.gradle.org/current/userguide/single_versions.html
  [4]: https://github.com/gradle/gradle/blob/57f994bf5a6388ea2883843c4e00b92b02491c98/subprojects/docs/src/docs/userguide/dep-man/02-declaring-dependency-versions/single_versions.adoc?plain=1#L45

**Reply by bric3** [reply]
Actually, I went back to [Gradle 6.9 doc][1], and in this case `SNAPSHOT` don't have specific meanings

> The string `SNAPSHOT` has **no special meaning**, and is sorted alphabetically like any other string part: `1.0-SNAPSHOT` < `1.0-alpha` < `1.0-zeta` < `1.0-rc` < `1.0`

This was changed starting from [Gradle 7.0][2]

> The string `snapshot` will be ordered higher than `rc`: `1.0-RC` < `1.0-SNAPSHOT` < `1.0`.

[1]: https://github.com/gradle/gradle/blob/a6198e44749b18b37e26b3b3467db17e034bcff4/subprojects/docs/src/docs/userguide/dep-man/02-declaring-dependency-versions/single_versions.adoc?plain=1#L57
[2]: https://github.com/gradle/gradle/blob/d5661e3f0e07a8caff705f1badf79fb5df8022c4/subprojects/docs/src/docs/userguide/dep-man/02-declaring-dependency-versions/single_versions.adoc?plain=1#L45

**Reply by JavierSegoviaCordoba** [reply]
You are right but it is weird the current docs still points to an old behavior. I am going to report that in the Gradle Slack.

To be sure, I will do some tests with dynamic versioning and Renovate (to report on that tool this change), and I will remove that behavior.

Thank you for your help and feedback!

**Reply by JavierSegoviaCordoba** [reply]
@bric3 I did some tests and Renovate fixed this behavior too, so I will try to update it ASAP.

Issue https://github.com/JavierSegoviaCordoba/semver-gradle-plugin/issues/92

**Reply by bric3** [reply]
Maybe having the property might be interesting for consumers that still require an old version of Gradle, but this may be not worth it as time passes and gradle is already pushing the first RC of 8.0.

**Reply by bric3** [reply]
Also actually in #92 it's starting from 7.0, not 6.9 ;)

**Reply by JavierSegoviaCordoba** [reply]
@bric3 I am not a fan of keeping compatibility with old versions except if it is a "real bug" in my plugin that should be backported, even more here as I am not sure if Renovate keeps different ordering for each Gradle version.

They can stick to an old version of my plugin for that until they can upgrade Gradle.

**Reply by JavierSegoviaCordoba** [reply]
Thank you again for your investigation, I added a comment to simplify the conclusion of our discussion to mark it as the answer.

**Reply by bric3** [reply]
you're welcome :)
More thanks for the plugin !

**Comment by JavierSegoviaCordoba** [comment, marked as answer]
As @bric3 mentioned this behavior changed on 7.0+ and it is fixed for the Semver plugin here https://github.com/JavierSegoviaCordoba/semver-gradle-plugin/issues/92

> Typically [this point](https://github.com/gradle/gradle/blob/57f994bf5a6388ea2883843c4e00b92b02491c98/subprojects/docs/src/docs/userguide/dep-man/02-declaring-dependency-versions/single_versions.adoc?plain=1#L45) lead me to believe snapshots are actually higher than RCs
> 
> 
> > The strings rc, snapshot, final, ga, release and sp are considered higher than any other string part (sorted in this order): 1.0-zeta < 1.0-rc < 1.0-snapshot < 1.0-final < 1.0-ga < 1.0-release < 1.0-sp < 1.0.
